sui_programmability/examples/defi/sources/pool.move

```diff
@@ -0,0 +1,472 @@
+    public fun swap_sui<P, T>(
+        pool: &mut Pool<P, T>, sui: Coin<SUI>, ctx: &mut TxContext
+    ): Coin<T> {
+        assert!(coin::value(&sui) > 0, EZeroAmount);
+
+        let sui_balance = coin::into_balance(sui);
+
+        // Calculate the output amount - fee
+        let (sui_reserve, token_reserve, _) = get_amounts(pool);
+        let output_amount = get_input_price(
+            balance::value(&sui_balance),
+            sui_reserve,
+            token_reserve,
+            pool.fee_percent
+        );
+
+        balance::join(&mut pool.sui, sui_balance);
+        coin::withdraw(&mut pool.token, output_amount, ctx)
+    }
+
+    /// Entry point for the `swap_token` method. Sends swapped SUI
+    /// to the sender.
+    entry fun swap_token_<P, T>(
+        pool: &mut Pool<P, T>, token: Coin<T>, ctx: &mut TxContext
+    ) {
+        transfer::transfer(
+            swap_token(pool, token, ctx),
+            tx_context::sender(ctx)
+        )
+    }
+
+    /// Swap `Coin<T>` for the `Coin<SUI>`.
+    /// Returns the swapped `Coin<SUI>`.
+    public fun swap_token<P, T>(
+        pool: &mut Pool<P, T>, token: Coin<T>, ctx: &mut TxContext
+    ): Coin<SUI> {
+        assert!(coin::value(&token) > 0, EZeroAmount);
+
+        let tok_balance = coin::into_balance(token);
+        let (sui_reserve, token_reserve, _) = get_amounts(pool);
+        let output_amount = get_input_price(
+            balance::value(&tok_balance),
+            token_reserve,
+            sui_reserve,
+            pool.fee_percent
+        );
+
+        assert!(sui_reserve > 0 && token_reserve > 0, EReservesEmpty);
+
+        balance::join(&mut pool.token, tok_balance);
+        coin::withdraw(&mut pool.sui, output_amount, ctx)
+    }
+
+    /// Entrypoint for the `add_liquidity` method. Sends `Coin<LSP>` to
+    /// the transaction sender.
+    entry fun add_liquidity_<P, T>(
+        pool: &mut Pool<P, T>, sui: Coin<SUI>, token: Coin<T>, ctx: &mut TxContext
+    ) {
+        transfer::transfer(
+            add_liquidity(pool, sui, token, ctx),
+            tx_context::sender(ctx)
+        );
+    }
+
+    /// Add liquidity to the `Pool`. Sender needs to provide both
+    /// `Coin<SUI>` and `Coin<T>`, and in exchange he gets `Coin<LSP>` -
+    /// liquidity provider tokens.
+    public fun add_liquidity<P, T>(
+        pool: &mut Pool<P, T>, sui: Coin<SUI>, token: Coin<T>, ctx: &mut TxContext
+    ): Coin<LSP<P, T>> {
+        assert!(coin::value(&sui) > 0, EZeroAmount);
+        assert!(coin::value(&token) > 0, EZeroAmount);
+
+        let sui_balance = coin::into_balance(sui);
+        let token_balance = coin::into_balance(token);
+
+        let (sui_amount, _, lsp_supply) = get_amounts(pool);
+
+        let sui_added = balance::value(&sui_balance);
+        let share_minted = (sui_added * lsp_supply) / sui_amount;
+
+        let sui_amt = balance::join(&mut pool.sui, sui_balance);
+        let tok_amt = balance::join(&mut pool.token, token_balance);
```

Comment:
Seems like `token_balance` should be read somewhere in this function and checked against the return value of `get_amoints`. E.g.

```
add_liquidity(ten_sui, ten_coins)
add_liquidity(ten_sui, zero_coins)
```

will give the caller the same number of LSP tokens, which doesn't seem right.

Reply:
Agreed I'll add a check that number of T is higher than the expected value.